test/shmemx/mt_perf_counter.c

```diff
@@ -0,0 +1,171 @@
+/*
+ *  Copyright (c) 2018 Intel Corporation. All rights reserved.
+ *  This software is available to you under the BSD license below:
+ *
+ *      Redistribution and use in source and binary forms, with or
+ *      without modification, are permitted provided that the following
+ *      conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+/* Multi-threaded tests for validation of memory barrier implemented in 
+ * different synchronization routines.
+*/
+
+#include <stdio.h>
+#include <pthread.h>
+#include <shmemx.h>
+#include <unistd.h>
+
+/* For systems without the PThread barrier API (e.g. MacOS) */
+#include "pthread_barrier.h"
+
+#define T 2
+#define ITER 100
+#define WINDOW 64
+#define LENGTH 1024
+
+int me, npes;
+char *src_array, *dest_array;
+uint64_t c_put, c_get, p_put, p_get, target;
+shmem_ctx_t *active_contexts[T];
+int extra_ctx_count = 0;
+
+pthread_barrier_t fencebar;
+pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;
+
+static void add_ctx_to_collect(shmem_ctx_t *ctx) {
+    pthread_mutex_lock(&mutex);
+    active_contexts[extra_ctx_count++] = ctx;
+    pthread_mutex_unlock(&mutex);
+}
+
+static void * thread_main(void *arg) {
+    int tid = *(int *) arg;
+    int i, j;
+    int partner = ((npes % 2 == 0) ? (me % 2 == 0 ? me + 1 : me - 1) : 
+                                     (me % 2 != 0 ? me - 1 : 
+                                     (me == npes - 1) ? me : me + 1));
+
+    shmem_ctx_t ctx;
+    shmem_ctx_create(SHMEM_CTX_PRIVATE, &ctx);
+    add_ctx_to_collect(&ctx);
+
+    for (i = 0; i < ITER; i++) {
+        for (j = 0; j < WINDOW; j++) {
+            shmem_ctx_putmem_nbi(ctx, dest_array + tid * LENGTH, 
+                                 src_array + tid * LENGTH, LENGTH, partner);
+        }
+        shmem_ctx_quiet(ctx);
+    }
+
+    shmem_ctx_destroy(ctx);
```

Comment:
Will the collector access the `ctx` after it is destroyed?

Reply:
I was planning to add a remove method, but instead will invoke `shmemx_pcntr_*` from the creator thread itself.

Reply:
Do you want to just get rid of the collector thread in this unit test?  It can't observe much as long as all the activity is happening on private contexts.